yb-voyager/src/tgtdb/conn_pool.go

```diff
@@ -44,27 +45,94 @@ type ConnectionPool struct {
 	sync.Mutex
 	params                    *ConnectionParams
 	conns                     chan *pgx.Conn
+	idleConns                 chan *pgx.Conn
 	connIdToPreparedStmtCache map[uint32]map[string]bool // cache list of prepared statements per connection
 	nextUriIndex              int
 	disableThrottling         bool
+	size                      int // current size of the pool
+	pendingConnsToClose       int
+	pendingConnsToCloseLock   sync.Mutex
 }
 
 func NewConnectionPool(params *ConnectionParams) *ConnectionPool {
 	pool := &ConnectionPool{
 		params:                    params,
-		conns:                     make(chan *pgx.Conn, params.NumConnections),
-		connIdToPreparedStmtCache: make(map[uint32]map[string]bool, params.NumConnections),
+		conns:                     make(chan *pgx.Conn, params.NumMaxConnections),
+		idleConns:                 make(chan *pgx.Conn, params.NumMaxConnections),
+		connIdToPreparedStmtCache: make(map[uint32]map[string]bool, params.NumMaxConnections),
 		disableThrottling:         false,
+		size:                      params.NumConnections,
+		pendingConnsToClose:       0,
+	}
+	for i := 0; i < params.NumMaxConnections; i++ {
+		pool.idleConns <- nil
 	}
 	for i := 0; i < params.NumConnections; i++ {
-		pool.conns <- nil
+		c := <-pool.idleConns
+		pool.conns <- c
 	}
 	if pool.params.SessionInitScript == nil {
 		pool.params.SessionInitScript = defaultSessionVars
 	}
 	return pool
 }
 
+func (pool *ConnectionPool) GetNumConnections() int {
+	return pool.size
+}
+
+func (pool *ConnectionPool) UpdateNumConnections(delta int) error {
+	pool.pendingConnsToCloseLock.Lock()
+	defer pool.pendingConnsToCloseLock.Unlock()
+
+	newSize := pool.size + delta
+
+	if newSize < 1 || newSize > pool.params.NumMaxConnections {
+		return fmt.Errorf("invalid new pool size %d. "+
+			"Must be between 1 and %d", newSize, pool.params.NumMaxConnections)
+	}
+
+	if delta == 0 {
+		log.Infof("adaptive: No change in pool size. Current size is %d", pool.size)
+		return nil
+	}
+	if delta > 0 {
+		// for increases, process them synchronously.
+		// If there is non-zero pendingConnsToClose, then we reduce that count.
+		if pool.pendingConnsToClose >= 0 {
+			// since we are increasing, we can just reduce the pending close count
+			if pool.pendingConnsToClose > delta {
+				log.Infof("adaptive: Decreasing pendingConnsToClose by %d", delta)
+				pool.pendingConnsToClose -= delta
+				delta = 0
+			} else {
+				log.Infof("adaptive: Decreasing pendingConnsToClose to 0 by %d", pool.pendingConnsToClose)
+				delta -= pool.pendingConnsToClose
+				pool.pendingConnsToClose = 0
+			}
+		}
```

Comment:
```
		if pool.pendingConnsToClose > 0 {
			// since we are increasing, we can just reduce the pending close count
			connCount := Math.min(pool.pendingConnsToClose, delta)
			delta -= connCount
			pool.pendingConnsToClose -= connCount
		}
```

Reply:
I find this a bit confusing to be honest. 😅 
Even though an if-else block is more verbose, I prefer it because it's simpler to reason.

Reply:
1. Zero equality in the condition is still of no use.
2. I dont think it will be confusing or hard to read. I feel it was quite verbose, You can add a comment saying - `take as much connections as possible from pool`

Reply:
I think this is subjective. However, on second thoughts, it seems alright. I've made changes!